### PR TITLE
[NVPTX] Improved support for grid_constant

### PR DESCRIPTION
**Pull request by akshayrdeodhar**
- Supports escaped grid_constant pointers less conservatively. Casts uses inside Calls, PtrToInts, Stores where the pointer is a _value operand_ to generic address space, immediately before the escape, while keeping other uses in the param address space

- Related to: https://github.com/llvm/llvm-project/pull/96125

### CODE REVIEW

**Comment by akshayrdeodhar** [comment]
CC: @apaszke

File: llvm/lib/Target/NVPTX/NVPTXUtilities.cpp [hunk 1]
```diff
@@ -210,7 +210,8 @@ bool isParamGridConstant(const Value &V) {
   if (const Argument *Arg = dyn_cast<Argument>(&V)) {
     // "grid_constant" counts argument indices starting from 1
     if (Arg->hasByValAttr() &&
-        argHasNVVMAnnotation(*Arg, "grid_constant", /*StartArgIndexAtOne*/true)) {
+        argHasNVVMAnnotation(*Arg, "grid_constant",
+                             /*StartArgIndexAtOne*/ true)) {
```
Comment (jlebar) [inline llvm/lib/Target/NVPTX/NVPTXUtilities.cpp:214]: Nit: undo this nop change?

**Review by jlebar — APPROVED** [review]

File: llvm/lib/Target/NVPTX/NVPTXUtilities.cpp (hunk 1 above)
Comment (akshayrdeodhar) [reply llvm/lib/Target/NVPTX/NVPTXUtilities.cpp:214]: clang-format 😅 (was going above 80 chars in the previous MR)

**Comment by Artem-B** [comment]
@akshayrdeodhar , @jholewinski  -- is `cvta.param` intended to be used on *any* `.param` data? Can it be used in `__device__` functions if we know we never write to the pointer? Or is it intended to be used on kernels only?

NVCC appears to use it for the kernels only, but I can't tell if that's an inherent restriction on the instruction use, or if it just happens that NVCC enables it for  parameters with `__grid_constant__` argument only, which can be used only by the kernels.

[PTX specs](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cvta) do not seem to specify it explicitly. They do link to "kernel function parameters" when they mention `.param`, but other than that it reads as if the instruction is applicable to all `.param` space data.

**Pull request merged**